lib/firenest/replicated_state/server.ex

```diff
@@ -0,0 +1,288 @@
+          handler = Handler.local_delete(handler, [value])
+          {:reply, :ok, %{state | handler: handler, remote: remote}}
+      end
+    else
+      {:reply, {:error, :already_present}, state}
+    end
+  end
+
+  def handle_call({:update, key, pid, arg}, _from, state) do
+    %{store: store, handler: handler, remote: remote} = state
+
+    case Store.fetch(store, key, pid) do
+      {:ok, value, delta} ->
+        case Handler.local_update(handler, arg, key, pid, delta, value) do
+          {:put, value, delta, handler} ->
+            remote = Remote.local_update(remote, key, pid, value, delta)
+            store = Store.local_update(store, key, pid, value, delta)
+            {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+          {:delete, value, delta, handler} ->
+            remote = Remote.local_update(remote, key, pid, value, delta)
+
+            case Store.local_delete(store, key, pid) do
+              # The value returned from update is fresher
+              {:ok, _value, store} ->
+                remote = Remote.local_delete(remote, key, pid)
+                handler = Handler.local_delete(handler, [value])
+                {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+              {:last_member, _value, store} ->
+                unlink_flush(pid)
+                remote = Remote.local_delete(remote, key, pid)
+                handler = Handler.local_delete(handler, [value])
+                {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+            end
+        end
+
+      :error ->
+        {:reply, {:error, :not_present}, state}
+    end
+  end
+
+  def handle_call({:delete, key, pid}, _from, state) do
+    %{store: store, handler: handler, remote: remote} = state
+
+    case Store.local_delete(store, key, pid) do
+      {:ok, value, store} ->
+        remote = Remote.local_delete(remote, key, pid)
+        handler = Handler.local_delete(handler, [value])
+        {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+      {:last_member, value, store} ->
+        unlink_flush(pid)
+        remote = Remote.local_delete(remote, key, pid)
+        handler = Handler.local_delete(handler, [value])
+        {:reply, :ok, %{state | store: store, handler: handler, remote: remote}}
+
+      :error ->
+        {:reply, {:error, :not_present}, state}
+    end
+  end
+
+  def handle_call({:delete, pid}, _from, state) do
+    %{store: store, handler: handler} = state
+
+    case Store.local_delete(store, pid) do
+      {:ok, deletes, store} ->
+        unlink_flush(pid)
+        handler = Handler.local_delete(handler, deletes)
+        # TODO: how do we handle remote in here?
+        {:reply, :ok, %{state | store: store, handler: handler}}
+
+      :error ->
+        {:reply, {:error, :not_member}, state}
+    end
+  end
+
+  def handle_call({:list, key}, _from, state) do
```

Comment:
Maybe we should do `handle_call(:store, _from, state)` instead of passing anonymous functions around?

Reply:
I did this because I didn't want to "leak" the internal state outside the process that explicitly. Also, the main `ReplicatedState` module does not know about the other internal modules, it only deals with the `Server` right now.

Reply:
Shall we return a MFArgs then? My only concern is passing anonymous functions around between processes. :)